### ISSUE TITLE
BeamMP Chat Filter

### ISSUE THREAD

**Issue by Holyguaka** [Starystars67/BeamNG-MP#402]
*Edited to remove the previous content and add criteria for implementation.*

Add a client-side filter for the chat, which each user can toggle on or off in their settings. The following word filters ship with the mod:

- Basic profanity filter (e.g. swearing)
- Hate speech filter (e.g. racism)
- Inappropriate content filter (e.g. sexually explicit content)

An additional multiline text field allows the user to specify words for a custom filter. All enabled filters are added together to form the filter actually used to filter chat.

- [x] Add the three preset filters with some basic wordlists
- [x] Implement settings page for turning these off and on (default OFF)
- [x] Custom filters textbox on the same settings page to allow users to specify words themselves
- [x] Replace offending words with characters like `****`

**Comment by lionkor** [comment]
A list like this is rarely a good solution. The impact of swearing is rarely tied to the words used, and I'm not sure this has been thought through. Someone trying to offend somebody will find the words, meanwhile people not trying to offend others will be hit by this. 

The proposed list includes "omg" and "lmao" among other common words, not sure thats necessary.

I agree its difficult to moderate swearing, but I also believe that swearing in itself isnt a problem. I dont think this will, at all, fix the issue, and just make it more difficult to report someone (because now they will call you a "fucking nword" instead of rhe slur... can I still report that?).

**Comment by lionkor** [comment]
Sorry misclick didnt mean to close.
Maybe a list comstrained to just slurs would do a better job, since it wouldnt catch people saying it legitimately?

**Comment by Starystars67** [comment]
I like the idea as it can be used to filter some of the more offending / unacceptable content, especially for the younger members of the community. Indeed some of this is OTT but some would be good to have. I think the idea here is to filter the worst of it for the users who want it when in open space.

**Comment by Holyguaka** [comment]
Here's an updated list. Didn't realise it was quite so PG. Think this is more realistic?

```
anal
anus
arse
ballsack
bitch
biatch
blowjob
blow job
bollock
bollok
boner
boob
bugger
butt
buttplug
clitoris
cock
coon
cunt
damn
dick
dildo
dyke
fag
feck
fellate
fellatio
felching
fuck
f u c k
flange
homo
jizz
knobend
knob end
labia
muff
nigger
nigga
omg
penis
prick
pube
pussy
queer
scrotum
sex
shit
s hit
sh1t
slut
smegma
spunk
tit
twat
vagina
wank
whore
```

**Comment by Kronki-705** [comment]
You know there is technically the N word in the Discord now?

**Comment by lionkor** [comment]
BeamMP is just risking to get people not to use chat very much, I don't think this will make a difference. What happens when someone says `stop ramming you **** ****** ****` and `**** you`? Is that no longer reportable?

This needs way way way more elaboration on how it'll work, how these reports would be handled, etc. IMO as a player.

**Comment by Starystars67** [comment]
The idea of this was discussed as a toggleable setting client side so that users who wish to not see such content can have the choice to disable it.

I was thinking that maybe this could be broken down into categories E.g. Racial, Profanity, Homophobic to name a few.

This also gives carers and guardians the ability to at least have the choice to protect their children from such content too.

**Comment by Starystars67** [comment]
> BeamMP is just risking to get people not to use chat very much, I don't think this will make a difference. What happens when someone says `stop ramming you **** ****** ****` and `**** you`? Is that no longer reportable?
> 
> This needs way way way more elaboration on how it'll work, how these reports would be handled, etc. IMO as a player.

The senders message would be fully visible to anybody who does not have the filter turned on. As such anyone inclined to report it still can.

I feel this is also a good opportunity to suggest in-game reporting features again as this would allow the collection of the message and the users ID at the time of occurrence.

With regards to further elaboration / discussion please feel free to contact me as this topic was discussed during the weekly staff meeting.

**Comment by lionkor** [comment]
Ah, okay, thank you! Wouldn't hurt to include all this info when making an issue, as I know I'm not the only player who looks at the issues and doesn't attend the meetings. No more questions from me tho :)

**Comment by Starystars67** [comment]
> Ah, okay, thank you! Wouldn't hurt to include all this info when making an issue, as I know I'm not the only player who looks at the issues and doesn't attend the meetings. No more questions from me tho :)

Indeed, I think issue is experience in writing issues/feature requests for new members really. Maybe we can look at refining/improving the templates for future users.

**Comment by Starystars67** [comment]
The requirements of this issue have been addressed as part of the UI work I am doing on the linked PR which is currently a draft until I am done.
